### PR TITLE
Test script  for updateRecommendation API

### PR DESCRIPTION
**Pull request by msvinaykumar**
added updateREcomendation API to scale test and enabled simultaneous uploading of results and update recommendations. Additionally, we have introduced extra flags to selectively post results or update recommendations only. Moreover, it is now possible to specify a start date as well.

syntax ex:

python3 -u quickTestScalability.py --ip master-0.kruizevin.lab.psi.pnq2.redhat.com  --port 30792 --count 1,1440 --minutesjump=15 --generaterecommendation  --parallelresultcount 1 --startdate=2023-01-01T00:00:00.000Z --name temp24Hrs3 --postresults

above example creates experiment name temp24Hrs3_1
with 15 days of result data also triggers updateRecommendations api  

Signed-off-by: msvinaykumar <vinakuma@redhat.com>

### CODE REVIEW

**Comment by chandrams** [comment]
@msvinaykumar - Please include functional test for updateRecommendations, I have listed a few scenarios below add if anything is missing to it.

Test scenarios for updateRecommendations, invoke updateRecommendations as below:

Post a valid experiment name & valid end time and validate recommendations are as expected (What is the valid range for end time?)
Post a valid experiment name, valid end time and valid start time (valid range), validate recommendations are as expected
Post the same experiment name & end time twice, validate recommendations - Should be same or within acceptable threshold values?

Validate status code & expected error messages for the below:
Post without any parameters
Post an invalid experiment name
Post only valid interval end time
Post a valid experiment name but an invalid end time 
Post a valid experiment name and an invalid end time format
Post a valid experiment name and an valid end time format but an invalid start time format
Post a valid experiment name, valid end time and invalid start time (future date or past date beyond accepted valid range)
Test for boundary values for the timestamps if there are any ranges

File: tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
+import pytest
+from helpers.fixtures import *
+from helpers.kruize import *
+from helpers.list_reco_json_validate import *
+from helpers.utils import *
+
+
+@pytest.mark.sanity
+def test_update_valid_recommendations_after_results_after_create_exp(cluster_type):
+    input_json_file = "../json_files/create_exp.json"
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:10]: Include description for all the tests included with expected behavior

File: tests/scripts/remote_monitoring_tests/rest_apis/test_e2e_workflow.py [hunk 2]
```diff
@@ -81,32 +80,26 @@ def test_list_recommendations_multiple_exps_from_diff_json_files(cluster_type):
 
             # Expecting that we have recommendations
             if j > 96:
-                # Sleep for 1 sec to get recommendations
-                time.sleep(1)
-
+                response = update_recommendations(experiment_name, interval_start_time, end_time)
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/rest_apis/test_e2e_workflow.py:83]: Please update all the tests in test_list_recommendations.py to invoke update_recommendations

File: tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        assert response.status_code == SUCCESS_STATUS_CODE
+
+
+@pytest.mark.negative
+def test_update_recommendations_without_experiment_name_end_time(cluster_type):
+    form_kruize_url(cluster_type)
+    response = update_recommendations(None, None, None)
+    data = response.json()
+    assert response.status_code == ERROR_STATUS_CODE
+    assert data['message'] == UPDATE_RECOMMENDATIONS_MANDATORY_DEFAULT_MESSAGE
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:240]: Shouldn't the error messages returned for negative tests match what is mentioned in the monitoringAPI.md ?#854

File: tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+
+@pytest.mark.negative
+def test_update_recommendations_without_end_time(cluster_type):
+    form_kruize_url(cluster_type)
+    experiment_name = "test123"
+    response = update_recommendations(experiment_name, None, None)
+    data = response.json()
+    assert response.status_code == ERROR_STATUS_CODE
+    assert data['message'] == UPDATE_RECOMMENDATIONS_MANDATORY_INTERVAL_END_DATE
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:250]: Here the experiment_name doesn't exist I think, will it still return error message related to interval end date? Would it return experiment_name is invalid ?

File: tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+@pytest.mark.negative
+def test_update_recommendations_with_end_time_precede_start_time(cluster_type):
+    form_kruize_url(cluster_type)
+    experiment_name = "test123"
+    start_time = "2023-01-03T00:15:00.000Z"
+    end_time = "2023-01-02T00:15:00.000Z"
+    response = update_recommendations(experiment_name, start_time, end_time)
+    data = response.json()
+    assert response.status_code == ERROR_STATUS_CODE
+    assert data['message'] == UPDATE_RECOMMENDATIONS_START_TIME_PRECEDE_END_TIME
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:273]: Please update the API doc with this error message

Comment (chandrams) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:273]: Can you please share the results for all these tests.

File: tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+                    # Check if duration
+                    assert INFO_DURATION_BASED_RECOMMENDATIONS_AVAILABLE_CODE in high_level_notifications
+                    data_section = recommendation_section["data"]
+                    short_term_recommendation = data_section[str(end_time)]["duration_based"]["short_term"]
+                    short_term_notifications = short_term_recommendation["notifications"]
+                    for notification in short_term_notifications.values():
+                        assert notification["type"] != "error"
+
+        response = update_recommendations(experiment_name, interval_start_time, end_time)
+        data = response.json()
```
[the remark this answers is not in the document]
Comment (msvinaykumar) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:102]: done

File: tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py (hunk 1 above)
Comment (msvinaykumar) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:16]: done

File: tests/scripts/remote_monitoring_tests/rest_apis/test_e2e_workflow.py (hunk 2 above)
Comment (msvinaykumar) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_e2e_workflow.py]: done

File: tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py (hunk 3 above)
Comment (msvinaykumar) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:267]: Update accordingly. In general it will display 'data not found'.

Comment (msvinaykumar) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:267]: done

File: tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py (hunk 4 above)
Comment (msvinaykumar) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:280]: Update accordingly. In general it will display 'data not found'.

File: tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py (hunk 5 above)
Comment (msvinaykumar) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:323]: Done

**Comment by msvinaykumar** [comment]
https://drive.google.com/drive/folders/17e1KE3eZKi8Waf3o7uvlMQ1SqJC8cpfW?usp=sharing

File: tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py (hunk 5 above)
Comment (msvinaykumar) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_update_recommendations.py:323]: > Can you please share the results for all these tests.

https://drive.google.com/drive/folders/17e1KE3eZKi8Waf3o7uvlMQ1SqJC8cpfW?usp=sharing

File: tests/scripts/remote_monitoring_tests/rest_apis/test_update_results.py [hunk 7]
```diff
@@ -507,11 +531,16 @@ def test_update_results_with_valid_and_invalid_interval_duration(test_name, inte
         assert response.status_code == SUCCESS_STATUS_CODE
         assert data['status'] == SUCCESS_STATUS
         assert data['message'] == UPDATE_RESULTS_SUCCESS_MSG
+    elif test_name == "invalid_zero_diff":
+        assert response.status_code == ERROR_STATUS_CODE
+        assert data['status'] == ERROR_STATUS
+        assert data['message'] == 'Out of a total of 1 records, 1 failed to save'
+        assert data['data'][0]['errorReasons'][0] == UPDATE_RESULTS_DATE_PRECEDE_ERROR_MSG
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/rest_apis/test_update_results.py:538]: UpdateResults.md needs to be updated with the errorReasons and error messages

File: tests/scripts/quickTestScalability.py [hunk 8]
```diff
@@ -18,359 +18,511 @@
 import datetime
 import requests
 import argparse
+import multiprocessing
```
Comment (chandrams) [inline tests/scripts/quickTestScalability.py:21]: @msvinaykumar - As discussed, please raise a separate PR for this on top of PR 832

**Comment by chandrams** [comment]
@msvinaykumar  - Tests did not run completely, extended and negative were not run due to the below error, did it work for you? Looking into this

Cmd used:

```
./test_autotune.sh -c minikube -i quay.io/vinakuma/autotune_operator:ros4 --testsuite=remote_monitoring_tests --resultsdir=/home/csubrama/kruize-scalability-res
```

Error:

```
FAILED test_create_experiment.py::test_create_exp_with_both_deployment_name_selector
FAILED test_create_experiment.py::test_create_exp_with_invalid_header - asser...
=========== 2 failed, 37 passed, 216 deselected in 82.78s (0:01:22) ============

/home/csubrama/github/vinay/autotune/tests/scripts/remote_monitoring_tests/remote_monitoring_tests.sh: line 127: 0 + 2
3
3
3
3
3
3
1
1
1
1: syntax error in expression (error token is "3
3
3
3
3
3
1
1
1
1")
```

**Comment by chandrams** [comment]
> @msvinaykumar - Tests did not run completely, extended and negative were not run due to the below error, did it work for you? Looking into this
> 
> Cmd used:
> 
> ```
> ./test_autotune.sh -c minikube -i quay.io/vinakuma/autotune_operator:ros4 --testsuite=remote_monitoring_tests --resultsdir=/home/csubrama/kruize-scalability-res
> ```
> 
> Error:
> 
> ```
> FAILED test_create_experiment.py::test_create_exp_with_both_deployment_name_selector
> FAILED test_create_experiment.py::test_create_exp_with_invalid_header - asser...
> =========== 2 failed, 37 passed, 216 deselected in 82.78s (0:01:22) ============
> 
> /home/csubrama/github/vinay/autotune/tests/scripts/remote_monitoring_tests/remote_monitoring_tests.sh: line 127: 0 + 2
> 3
> 3
> 3
> 3
> 3
> 3
> 1
> 1
> 1
> 1: syntax error in expression (error token is "3
> 3
> 3
> 3
> 3
> 3
> 1
> 1
> 1
> 1")
> ```

grep at line 124 seems to be matching these prints in the test output - Out of a total of 1 records, 1 failed to save
Please modify this to match the failed results count

File: tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py [hunk 9]
```diff
@@ -651,6 +708,7 @@ def test_list_recommendations_multiple_exps_with_missing_metrics(cluster_type):
         response = delete_experiment(create_exp_json_file)
         print("delete exp = ", response.status_code)
 
+
 @pytest.mark.extended
 @pytest.mark.parametrize("latest", ["true", "false"])
 def test_list_recommendations_with_only_latest(latest, cluster_type):
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py:714]: @msvinaykumar - rest_apis/test_list_recommendations.py::test_list_recommendations_with_only_latest[false] this test failed with the below error, can you please check

```
      # Validate the json against the json schema
        errorMsg = validate_list_reco_json(list_reco_json, list_reco_json_schema)
>       assert errorMsg == ""
E       assert [<ValidationError: "'monitoring_start_time' is a required property">, <ValidationError: "'monitoring_end_time' is a re...Error: "'variation' is a required property">, <ValidationError: "'monitoring_start_time' is a required property">, ...] == ''

test_list_recommendations.py:805: AssertionError
```

**Comment by dinogun** [comment]
@msvinaykumar does this need a rebase as well?

**Comment by chandrams** [comment]
@msvinaykumar - Please check the timestamp format in the error message

{"message":"Out of a total of 1 records, 1 failed to save","httpcode":400,"documentationLink":"","status":"ERROR","data":[{"interval_start_time":"Aug 17, 2023, 4:35:46 PM","interval_end_time":"Aug 17, 2023, 4:50:46 PM",

**Review by dinogun — APPROVED** [review]
LGTM
